File: easy/happy_numbers.c
#include <stdio.h>
#include <stdlib.h>

int happy(int a)
{
	int i = 0, ret = 0;
	char line[13];

	sprintf(line, "%d", a);
	do {
		int c = line[i] - 48;
		ret += c * c;
	} while (line[++i] != '\0');
	return ret;
}

int main(int argc, char *argv[])
{
    FILE *fp;
    char line[22];

    fp = fopen(*++argv, "r");
    while (fgets(line, 22, fp) != 0) {
	int i = 0, a = atoi(line), b[128] = {a,0};

	while (a != 1) {
	    int j = 0;
	    if (i > 126)
		goto out;
	    a = happy(a);
	    while (b[j] != 0) {
		if (b[j++] == a)
		    goto out;
	    }
	    b[++i] = a;
	}
	printf("1\n");
	continue;
out:
	printf("0\n");
    }
    return 0;
}

File: easy/happy_numbers.lua
function happy (a)
  local ret = 0
  while #a > 0 do
    local b = tonumber(a:sub(#a))
    ret = ret +  b * b
    a = a:sub(1, #a-1)
  end
  return tostring(ret)
end

for line in io.lines(arg[1]) do
  local a, b = line, {line}
  for _ = 1, 127 do
    if a == "1" or a == "0" then
      break
    end
    a = happy(a)
    for _, j in ipairs(b) do
      if j == a then
        a = "0"
        break
      end
    end
    b[#b + 1] = a
  end
  if a == "1" then
    print(1)
  else
    print(0)
  end
end
